### PR TITLE
feat(statusbar): detect if meetingbar is not displayed

### PR DESCRIPTION
**Pull request by jgoldhammer**
### Status
**READY**

### Description
- we now detect if the meetingbar icon is not displayed in the status bar
- we will now send a notification if meetingbar is not dispayed in the status bar due to low space
- the user can deactivate the notification in the preferences dialog or directly in the notification

image

image

## Checklist
- [x] Localized
- [x] Added to changelog:
  - [x] [Changelog View](https://github.com/leits/MeetingBar/blob/master/MeetingBar/Views/Changelog/Changelog.swift)
  - [x] [CHANGELOG.md](https://github.com/leits/MeetingBar/blob/master/CHANGELOG.md)

### Steps to Test or Reproduce

- open many apps with statusbar icons 
- open meetingbar at least, so it is not displayed (try to configure the displayed length of the title in status bar to the maximum)
- notice the notification

### CODE REVIEW

**Comment by jgoldhammer** [comment]
The method seems experimental. If mac is awake from sleep, the code detects that the statusbar item is hidden although the statusbar shows menubar...

**Comment by leits** [comment]
Looks good!
I once spent several hours trying to make a detector, but nothing came of it.

Does this method only work on wakeup or on open the app with lots of options in the menu bar also?

**Comment by jgoldhammer** [comment]
> Looks good! I once spent several hours trying to make a detector, but nothing came of it.
> 
> Does this method only work on wakeup or on open the app with lots of options in the menu bar also?

It will constantly detect every change in the menubar also on startup of meetingbar.

https://github.com/leits/MeetingBar/pull/435/files#diff-930b4ca4efafa0cef4455a6e6f49bdb7016c5a1d88cd1775bcfd17cbb3142513R63-R73

**Review by 0bmxa — COMMENTED** [review]
Very nice feature! I've been personally missing this! 
I tested it briefly, and it seemed to work well.

**Copy**
I left a few suggestions, regarding the copy text, mostly to make it a bit shorter or to align spelling/naming with Apple's guidelines (e.g. [to call it Menu Bar](https://developer.apple.com/design/human-interface-guidelines/macos/extensions/menu-bar-extras/) on macOS), but really nothing important, so feel free to ignore them.

**UX**
From a UX perspective I'd expect such a feature (in the future) to actually "take care of the problem", i.e. make it visible again (which probably means shortening/hiding the title), but this doesn't have to be in the current version.
But, I'd enable the setting by default, as it's very helpful for users who experience this issue, and shouldn't bother anybody who doesn't.

Thanks a lot for this!

**Comment by jgoldhammer** [comment]
I have made the necessary changes - please review again @leits ...

File: MeetingBar/Notifications.swift [hunk 1]
```diff
@@ -48,8 +63,19 @@ func sendUserNotification(_ title: String, _ text: String) {
     let content = UNMutableNotificationContent()
     content.title = title
     content.body = text
+    if categoryIdentier != nil {
```
[the remark this answers is not in the document]
Comment (leits) [reply MeetingBar/Notifications.swift:91]: It doesn't look like that's changed. Maybe you forgot to push these changes?

**Comment by leits** [comment]
@jgoldhammer, the notification is also sent when you open any application in full screen.
I spent a lot of time finding a way to check if the status bar is visible, but couldn't find any.

Do you also have this problem?

**Comment by jgoldhammer** [comment]
Hmm, indeed this is a problem I have not tested before. I also could not find any way to detected if any app is in fullscreen or the status bar is hidden...

**Comment by jgoldhammer** [comment]
It seems that this is only reliable available with access to the accessibility api on mac - see alt-tab for an example 

https://github.com/lwouis/alt-tab-macos/blob/0328538b3aba993ed4c6cbdae138f87179886eb9/src/logic/events/AccessibilityEvents.swift

**Comment by jgoldhammer** [comment]
I am now having a solution to detect if there is any app in fullscreen mode and do not trigger the notification when an app enters fullscreen.
The downside is that meetingbar needs the screen recording permission under macos. I am little bit frustrated about this situation because macOS does not allow to get the information if there is any app in fullscreen without that permission. 

image

You can read more here:
https://stackoverflow.com/a/66403010

What do you think, @leits, how to proceed here?

**Comment by leits** [comment]
This is definitely not the permission that we want to ask the user.

Thank you for your research Jens!
But let's suspend this feature and keep an eye on macOS updates.

**Pull request merged**